Adding Glossary (#170807)
pytorch/pytorch: docs/source/_glossary.md

(glossary)=
# PyTorch Glossary

This glossary provides definitions for terms commonly used in PyTorch documentation.

```{glossary}

ATen
   Short for "A Tensor Library". The foundational tensor and mathematical
   operation library on which all else is built.

Operation
   A unit of work. For example, the work of matrix multiplication is an operation
   called `aten::matmul`.

Native Operation
   An operation that comes natively with PyTorch ATen, for example `aten::matmul`.

Custom Operation
   An Operation that is defined by users and is usually a {term}`Compound Operation`.
   For example, this [tutorial](https://docs.pytorch.org/tutorials/advanced/custom_ops_landing_page.html)
   details how to create Custom Operations.

Kernel
    Refers to the "on device" implementation of a given operation. Reffering to the part of the Operation that actually computes the result.

Composite Operation

>    A Compound Operation is composed of other operations. Its kernel is usually

Author: ```suggestion
   A Composite Operation is composed of other operations. Its kernel is usually
```

(glossary)=
# PyTorch Glossary

This glossary provides definitions for terms commonly used in PyTorch documentation.

```{glossary}

ATen
   Short for "A Tensor Library". The foundational tensor and mathematical
   operation library on which all else is built.

Operation
   A unit of work. For example, the work of matrix multiplication is an operation
   called `aten::matmul`.

Native Operation
   An operation that comes natively with PyTorch ATen, for example `aten::matmul`.

Custom Operation
   An Operation that is defined by users and is usually a {term}`Compound Operation`.

>    For example, this [tutorial](https://pytorch.org/docs/stable/notes/extending.html)

Author: ```suggestion
   For example, this [tutorial](https://docs.pytorch.org/tutorials/advanced/custom_ops_landing_page.html)
```

   details how to create Custom Operations.

Kernel

>    Implementation of a PyTorch operation, specifying what should be done when an

Author: ```suggestion
    Refers to the "on device" implementation of a given operation. Reffering to the part of the Operation that actually computes the result.
```

>    operation executes.

Author: ```suggestion

```

> Compound Operation

Author: ```suggestion
Composite Operation
```

   A Compound Operation is composed of other operations. Its kernel is usually
   device-agnostic. There are two variants of this: Composite Implicit Autograd, where no
   autograd formula is required as it is derived implicitly from the operations composing this
   one. And Composite Explicit Autograd, where there is an explicit autograd formula.

Composite Operation
   Same as {term}`Compound Operation`.

Non-Leaf Operation
   Same as {term}`Compound Operation`.

> Leaf Operation

Author: ```suggestion
```

>    An operation that's considered a basic operation, as opposed to a {term}`Compound

Author: ```suggestion
```

>    Operation`. Leaf Operation always has dispatch functions defined, usually has a

Author: ```suggestion
```

>    derivative function defined as well.

Author: ```suggestion
```

Device Kernel
   Device-specific kernel of a {term}`Leaf Operation`.

> Compound Kernel

Author: ```suggestion
```

>    Opposed to {term}`Device Kernels<Device Kernel>`, Compound kernels are usually

Author: ```suggestion
```

>    device-agnostic and belong to {term}`Compound Operations<Compound Operation>`.

Author: ```suggestion
```

JIT
   Just-In-Time Compilation.

Tracing

>    Using `torch.jit.trace` on a function to get an executable that can be optimized

Author: ```suggestion
In PyTorch, tracing is a way to convert a PyTorch model (or function) into a static computation graph by running it once (or a few times) with example inputs and recording (“tracing”) the tensor operations that actually execute.```

>    using just-in-time compilation.

Author: ```suggestion
```
